[PATCH CATERPILLAR v3] Caterpillar mdev cxgb4

MykytaI opened the pull request:
This PR is basically cxgb4 driver + PR380. As soon as PR380 gets merged -- this PR will be rebased so that only one patch remains for the merge.

Bill-Fischofer-Linaro commented on platform/linux-generic/m4/odp_mdev.m4:
```diff
@@ -0,0 +1,34 @@
+##########################################################################
+# Enable mdev support
+##########################################################################
+mdev_support=no
```
Same comments as for PR #380, moreover we shouldn't duplicate commits in two PRs as that will cause merge conflicts. I suggest you reorg these into: 

1. A basic mdev PR that adds the common code. This can be reviewed and merged first.

2. Once that's in, individual PRs for the various mdev drivers become a lot simpler to add on top of this.

Bill-Fischofer-Linaro commented:
@muvarov I don't see a Travis run associated with this PR. Any idea why not?

heyi-linaro commented:
@MykytaI , I've merged PR380 would you like to rebase this PR and also check the CI and shippable problems?

MykytaI commented on platform/linux-generic/pktio/mdev/cxgb4.c:
```diff
@@ -0,0 +1,887 @@
+		if (odp_unlikely(pkt_len > CXGB4_TX_BUF_SIZE)) {
+			tx_pkts++;
+			continue;
+		}
```
Statistics are planned to be added at later stage.

MykytaI commented on platform/linux-generic/pktio/mdev/cxgb4.c:
```diff
@@ -0,0 +1,887 @@
+
+		if (odp_unlikely(rxq->offset + pkt_len > ODP_PAGE_SIZE)) {
+			/* TODO: reset the HW and reinit ? */
+			ODP_ABORT("Packet write beyond buffer boundary\n");
```
I would say that this condition should never ever happen. If it still happens -- it's a fatal error which requires system reboot because it's actually random memory corruption and you can't count on system behaving well after that.

Furthermore, with wonderful Chelsio HW it's impossible to reset it.

We can put cxgb4 pktio into error state permanently if you think it will be enough (with respect to memory corruption statement).

Bill-Fischofer-Linaro replied:
Yes, that way the application at least gets the opportunity to do something. Always best to let the application decide if termination is necessary.

MykytaI commented on platform/linux-generic/pktio/mdev/cxgb4.c:
```diff
@@ -0,0 +1,887 @@
+		}
+
+		pkt = odp_packet_alloc(pkt_cxgb4->pool, pkt_len);
+		if (odp_likely(pkt != ODP_PACKET_INVALID)) {
```
I've rearranged code to allocate packet before driver state is updated and break if allocation fails. The initial code was written with an assumption that code would be much more complex if we would try to revert driver state on packet allocation failure, but after some brainstorming code looks even simpler now (I hope).

Regarding odp_packet_alloc_multi: I believe this API is more of use to application developer because he/she already knows expected packet size. In drivers we don't know what will be length of RXed packets. Furthermore, depending on HW, we may or may not know how many packets we have to handle.

Even if the number of packets at hand is known (like in e1000e) -- it's still questionable if it's optimal to always allocate MTU-sized packets ... at least when we anyway copy packet payload. Of course all of this is rendered null and void as soon as we implement zerocopy.

MykytaI commented:
@Bill-Fischofer-Linaro , regarding OO approach: this was actually our intention to keep drivers as close to each other as possible. In fact, drivers most probably will need to implement only open/close and recv/send methods.

As mediated devices are supposed to be compatible with all linux standard iocts/netlink/etc APIs -- I think we can write generic 'linux' pktio methods and use those both in mdev and other pktios.

MykytaI commented:
@heyi-linaro , I've rebased the patch and addressed at least some of @Bill-Fischofer-Linaro comments.

The checkpatch failure seems to be false positive:
WARNING: Possible switch case/default not preceeded by break or fallthrough comment
#452: FILE: platform/linux-generic/pktio/mdev/cxgb4.c:420:
+       case VFIO_NET_MDEV_TX_RING:

The preceding case is clearly ending with 'return'.

heyi-linaro commented:
Merged and closed.
